A get-started tutorial (#24)
JuliaManifolds/LieGroups.jl: tutorials/getstarted.qmd

# 🚀 Get Started with LieGroups.jl

## Introduction

This tutorial introduces both a few basics of [Lie groups](https://en.wikipedia.org/wiki/Lie_group) as well as how to use [LieGroups.jl](https://juliamanifolds.github.io/LieGroups.jl/stable/). The Lie groups we consider are the rotations in the plane,
or [special orthogonal group](../groups/special_orthogonal_group.md) as well as rigid body motions,
or in other words the [special euclidean group](../groups/special_euclidean_group.md).

> In a nutshell a Lie group $𝒢$ is a [manifold](https://en.wikipedia.org/wiki/Manifold) $ℳ$ that also is a group a group operation.

Reviewer: ```suggestion
In a nutshell a Lie group $𝒢$ is a [manifold](https://en.wikipedia.org/wiki/Manifold) $ℳ$ that also is a group with a group operation.
```

> A manifold informally be described as a set that “locally looks like a Euclidean space”

Reviewer: ```suggestion
A manifold can informally be described as a set that “locally looks like a Euclidean space”
```

and has some way to measure angles at every point, to be precise an inner product, that is allowed to (smoothly)

> depend on the point we are at.

Reviewer: This isn't necessary for Lie groups, so I'd suggest skipping the inner product here.

Author: After introducing the left invariant vector fields and identifying everything with the Lie algebra, I do agree – but I do that in the following

A [group](https://en.wikipedia.org/wiki/Group_(mathematics)) means that on the set there also

> exists an operation, within this package usually denoted by $∘: 𝒢 × 𝒢 → 𝒢$$,

Reviewer: ```suggestion
exists an operation, within this package usually denoted by $∘: 𝒢 × 𝒢 → 𝒢$,
```

`LieGroups.jl` uses the manifolds defined in [`Manifolds.jl`](https://juliamanifolds.github.io/Manifolds.jl/stable/). For more details on that, see also the [introductory tutorial](https://juliamanifolds.github.io/Manifolds.jl/stable/tutorials/getstarted/) there.

> A Lie group is usually written as a tuple $(ℳ,∘)$ os the manifold and the group operation thereon.

Reviewer: ```suggestion
A Lie group is usually written as a tuple $(ℳ,∘)$ with the manifold and the group operation thereon.
```

For more theoretical background, see for example [HilgertNeeb:2012](@cite), especially their Chapter 9.

## Rotations on the plane $ℝ^2$ and in $ℝ^3$.

This first part considers rotations in the plane. These can be represented by rotation matrices
```math

> R_α = \begin{pmatrix} \cos(α),& -\sin(α)\\ \sin(α) & \cos(α) \end{pmatrix},\qquad α ∈ ℝ,

Reviewer: ```suggestion
R_α = \begin{pmatrix} \cos(α)& -\sin(α)\\ \sin(α) & \cos(α) \end{pmatrix},\qquad α ∈ ℝ,
```

```

where we already represent all possible rotations when restricting $α$ to $[0,2π]$.
Adding two matrices $R_α + R_β$ does not yield a valid rotation matrix again, but we
can have these matrices in mind as being “locally like the real line”, since changing the rotation angle
slightly yields rotation matrices that are “close by” in the sense that they perform nearly the same rotation
when applied to a point $x ∈ ℝ^2$ by computing $R_αx$.

> Matrix multiplication $R_β = R_{α_1}*R_{α_1}$ does yield a new rotation and also fulfils all other

Reviewer: ```suggestion
Matrix multiplication $R_β = R_{α_1} R_{α_2}$ does yield a new rotation and also fulfills all other
```
I'm not a fan of using `*` for multiplication in math.

Author: Indeed. That is not nice. We could add a `\cdot`? But we could also leave it out, sure.

Reviewer: Either `\cdot` or nothing, depending on your preference :slightly_smiling_face:

Author: Ok let‘s do nothing for now then. That should also be clear.

> properties to yield a group operation. Using trigonometric identities we observe that $β = α_1 + α_2$.

Reviewer: ```suggestion
properties to yield a group operation. Using trigonometric identities we observe that $β = α_1 + α_2 \pmod{2\pi}$.
```

Author: Formerly that is not necessary, since the rotation matrix does take care of that in cos/sin. 

I would prefer not writing a mod – but maybe adding that as a sentence that some might agree (have to think about a good phrasing).

Reviewer: Feel free to phrase it differently but I think it's important to make it more precise. We can either assume that $\beta=\alpha_1 + \alpha_2$ and conclude that the matrices multiply accordingly, or start from matrix multiplication (like you did) and conclude that we can only establish equality up to a multiple of $2\pi$.

Author: I will think about it and agree, it should be more precise and clear.

```

A first thing to do is to use [`is_point`](@ref) to verify they are valid.

```{julia}
(is_point(SO2, g), is_point(SO2, h))
```

The already mentioned group operation as the matrix multiplication, there is also a generic
function available, [`compose`](@ref) as well as its in-place variant [`compose!`](@ref).
The following two yield the same matrix

```{julia}
[ compose(SO2, g, h), g*h ]
```

Furthermore a lot of of functions are “passed down” tp the manifold, which is stored within
the [`LieGroup`](@ref). For example the dimension of the manifold, or the number of degrees of freedom
can be accessed via the ``[`manifold_dimension`](@extref `ManifoldsBase.manifold_dimension-Tuple{AbstractManifold}`)``{=commonmark}

```{julia}
manifold_dimension(SO2)
```

## The Lie algebra, and the Lie group exponential function.

For the following investigations, we consider the special orthogonal group $\mathrm{SO}(3)$,
that is rotations in a 3-dimensional space.
Besides one special rotation in the following code for completeness, a prominent, at first glace
maybe a bit “dull” point on this Lie group is the identity matrix `e`.

```{julia}
SO3 = SpecialOrthogonalGroup(3)

> g = [cos(π/3) -sin(π/3) 0.0; sin(π/3) cos(π/3) 0.0; 0.0 0.0 1.0]* [cos(π/4) 0.0 -sin(π/4); 0.0 1.0 0.0; sin(π/4) 0.0 cos(π/4)]

Reviewer: Maybe it would be more clear to use Rotations.jl here? You could just do something like `g = RotZ(π/3) * RotY(π/4)`.

Author: Sure, fair enough. Makes the code more readable. Will do that.

e = Matrix{Float64}(I,3,3)
g
```

> The element `g` can be seen as a rotation by $π/3$ in the x-y-plane combines with a rotation by $π/4$ in the x-z plane.

Reviewer: ```suggestion
The element `g` can be seen as a rotation by $π/3$ in the x-y-plane combined with a rotation by $π/4$ in the x-z plane.
```

The tangent space $T_e𝒢$ plays a special role and is called the [`LieAlgebra`](@ref) $𝔤$.

Similar to the ``[Riemannian exponential map](@extref `Base.exp-Tuple{AbstractManifold, Any, Any}`)``{=commonmark}
The exponential function $\exp_𝒢: 𝔤 → 𝒢$ maps Lie algebra tangent vectors $X ∈ T_e𝒢$

> to a point on the Lie group. This is implemented in [`exp(G::LieGroup, X)`](@ref). It's inverse

Reviewer: ```suggestion
to a point on the Lie group. This is implemented in [`exp(G::LieGroup, X)`](@ref). Its inverse
```

To “turn” the `X` into a vector with three elements, we need a basis of the tangent space
to decompose `X` into and “read off” its components.
While in general this can be done using the [bases of a tangent space](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/bases/) from [`ManifoldsBase.jl`](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/),
a very specific one is [`vee(::LieAlgebra, X)`](@ref)

```{julia}
c = vee(LieAlgebra(SO3), X)
```

and its inverse, [`hat(G::LieAlgebra, c)`](@ref)

```{julia}
X_ = hat(LieAlgebra(SO3),c)
```

## Representing tangent vectors

Consider the function of left group composition $λ_g(h) = g∘h$.
Its differential (or push forward) $Dλ_g(e)$ maps from $T_e𝒢$ to $T_g𝒢$
and is a linear bijection between both spaces.
Its adjoint differential (or pullback) $D^*λ_g(e)$ maps back from $T_g𝒢$ to $T_e𝒢$.

This allows to introduce an inner product on all tangent spaces, that smoothly varies with
the base point, we obtain the left-invariant metric

```math
⟨X,Y⟩_g = ⟨D^*λ_g(e)[X],D^*λ_g(e)⟩_e,
```

since it is invariant if you use the differential of the left composition to identify tangent vectors.

> We can even “spare” the pullbacks in this definition, when spring tangent vectors as their representants in the Lie algebra.

Reviewer: spring?

Author: ah autocorrect with too many typos. `representing`

> This is the default in `LieGroups.jl`.

Reviewer: Maybe it would be a good idea to add that this metric in general is not right-invariant, except for compact Lie groups and their direct products with `Translations`.

Author: Sure, maybe a bit of a technical note, since one has to say before also that right invariant can be build in a similar way. Will add that.

Author: Should we maybe add a reference for such a statement?

Reviewer: Sure. The most general statement of this kind that I know of is in
```julia
@article{LatifiToomanian:2013,
	title = {On the existence of bi-invariant {Finsler} metrics on {Lie} groups},
	volume = {7},
	issn = {2251-7456},
	doi = {10.1186/2251-7456-7-37},
	number = {1},
	journal = {Mathematical Sciences},
	author = {Latifi, Dariush and Toomanian, Megerdich},
	month = aug,
	year = {2013},
	pages = {37},
}
```
where they show that it's true not only for Riemannian metrics but also Finsler ones.

Author: Nice, will add that then. That it is even more general is fine with me

With respect to this metric on the manifold the exponential and logarithmic maps are given by

```math
\exp_g(X) = g ∘ \exp_𝒢(X)
\quad\text{with its inverse}\quad
\log_g(h) = \log_𝒢(g^{-1}∘h)
```

With respect to these geodesics and the representation in the Lie algebra, parallel transport

> simplifies to the identity. To still access the Riemannian exponential map with recpect to the

Reviewer: ```suggestion
simplifies to the identity. To still access the Riemannian exponential map with respect to the
```

> metric (compatible to the Levi-cevita connection) on the underlying Riemannian manifold, use

Reviewer: ```suggestion
metric (compatible to the Levi-Civita connection) on the underlying Riemannian manifold, use
```

For the following investigations, we consider the special orthogonal group $\mathrm{SO}(3)$,
that is rotations in a 3-dimensional space.
Besides one special rotation in the following code for completeness, a prominent, at first glace
maybe a bit “dull” point on this Lie group is the identity matrix `e`.

```{julia}
SO3 = SpecialOrthogonalGroup(3)
g = RotZ(π/3)*RotY(π/4)
e = Matrix{Float64}(I,3,3)
g
```
The element `g` can be seen as a rotation by $π/3$ in the x-y-plane combined with a rotation by $π/4$ in the x-z plane.

The tangent space $T_e𝒢$ plays a special role and is called the [`LieAlgebra`](@ref) $𝔤$.

Similar to the ``[Riemannian exponential map](@extref `Base.exp-Tuple{AbstractManifold, Any, Any}`)``{=commonmark}
The exponential function $\exp_𝒢: 𝔤 → 𝒢$ maps Lie algebra tangent vectors $X ∈ T_e𝒢$
to a point on the Lie group. This is implemented in [`exp(G::LieGroup, X)`](@ref). Its inverse
is the Lie group logarithmic function [`log(G::LieGroup, g)`](@ref).

```{julia}
X = log(SO3,g)
is_point(LieAlgebra(SO3), X)
```

```{julia}
exp(SO3,X)
```

The term “exponential” has at least three different meanings throughout Lie groups and Riemannian manifolds.

> To be precise, we call the just introduced one “exponential function”, since it often agrees with the [exponential](https://en.wikipedia.org/wiki/Exponential_function).

Reviewer: ```suggestion
To be precise, we call the just introduced one “exponential function”, since it often agrees with the [matrix exponential](https://en.wikipedia.org/wiki/Matrix_exponential).
```
It doesn't agree with the scalar exponential that often.

```
### Products and semidirect products of Lie groups

Similar to [product manifolds](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/metamanifolds/#ProductManifold)
(direct) [product Lie groups](../groups/product_group.md) are formed the same way. For two Lie groups
$𝒢 = 𝒢_1 × 𝒢_2$ is the product of the two manifolds together with the component wise application of the group operations:
the first group operation (of $𝒢_1$) to the first component and that of $𝒢_2$ to the second.

Instead of the component wise or “not interacting” variant of the (direct) product Lie groups,
for the [semidirect product Lie groups](../groups/semidirect_product_group.md) $𝒢 = 𝒢_1 ⋉ 𝒢_2$
we also require an action of how $𝒢_1$ acts on $𝒢_2$.
Semidirect here means that the first component of $𝒢$ is the same as for the direct product,
but before applying the second group operation on $𝒢_2$ one of the elements is “acted upon”
from an element of $𝒢_1$. The group operation reads for $(g_1,g_2), (h_1,h_2) ∈ 𝒢$

```math
(g_1,g_2) ∘ (h_1,h_2) := (g_1 ⋆ h_1, σ_{h_1}(g_2) ⋄ h_2).
```

### Rigid body motions

We obtain the [special Euclidean group](../groups/special_euclidean_group.md)

> $\mathrm{SE}(3) = \mathrm{SO}(3) ⋉ T(3)`` where the group action is the one discussed as an example before.

Reviewer: ```suggestion
$\mathrm{SE}(3) = \mathrm{SO}(3) ⋉ T(3)$ where the group action is the one discussed as an example before.
```

action is clear, because there exists a reasonable default. Otherwise the full form

```{julia}
LeftSemidirectProductLieGroup(SO3, T3, LeftGroupOperationAction())
```

is necessary. Here, the first `Left` for the semidirect product refers to the fact that
the left group acts on the right one before the right group operation is performed.
[`LeftGroupOperationAction`](@ref) refers to that the group operation – left multiplication with a matrix – is applied here.

For this case, any point $\mathrm{SE}(3)$ is a tuple of a rotation matrix and a vector.
We model this using [RecursiveArrayTools.jl](https://docs.sciml.ai/RecursiveArrayTools/stable/),
namely its `ArrayPartition`.
For example with the rotation matrix `g` from before we have

```{julia}
g1 = ArrayPartition(g, [1.0, 0.0, 2.0])
```

and for a pure translation we can reuse `e` as in

```{julia}
h1 = ArrayPartition(e, [0.0, 3.0, 0.0])
```

and we can summarize a few operations from before: both are valid points on `SE3`

```{julia}
(is_point(SE3, g1), is_point(SE3, h1))
```

we can perform the group operation

```{julia}
gh1 = compose(SE3, g1, h1)
```

apply the Lie group exponential

```{julia}
Y = log(SE3, gh1)
```

> and look at the manifold dimension, it is the sum of the single ones

Reviewer: ```suggestion
and look at the manifold dimension, it is the sum of dimensions of its components
```

```{julia}
manifold_dimension(SE3)
```

> and its coordinates as

Reviewer: ```suggestion
The coordinates of `Y` are
```
I think this sentence is too long.
